### PR TITLE
Add MetalKernels

### PR DESCRIPTION
**Pull request by vchuravy**
I would advocate for not having the backend packages in the Project.toml
they are all rather large and loading them unconditionally is a pain (and not supported on 1.6)

Instead we use Requires for 1.6-1.8 and extensions on 1.9

Also adds the experimental MetalKernels package + CI

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -17,23 +17,23 @@ using SafeTestsets, Test
 
 const GROUP = get(ENV, "GROUP", "CUDA")
 
-if GROUP == "CUDA"
-    @time @safetestset "EnsembleGPUArray" begin include("ensemblegpuarray.jl") end
-    @time @safetestset "EnsembleGPUArray OOP" begin include("ensemblegpuarray_oop.jl") end
-    @time @safetestset "EnsembleGPUArray SDE" begin include("ensemblegpuarray_sde.jl") end
-    @time @safetestset "EnsembleGPUArray Input Types" begin include("ensemblegpuarray_inputtypes.jl") end
+# if GROUP == "CUDA"
```
Comment (ChrisRackauckas) [inline test/runtests.jl:20]: ?

File: ext/MetalExt.jl [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module MetalExt
+    using KernelAbstractions
+    using Metal, MetalKernels, Adapt
+    import DiffEqGPU
+
+    DiffEqGPU.maxthreads(::MetalDevice) = 256
+    DiffEqGPU.maybe_prefer_blocks(::MetalDevice) = MetalDevice()
+
+    # TODO move to KA
+    Adapt.adapt_storage(::CPU, a::MtlArray) = adapt(Array, a)
```
Comment (utkarsh530) [inline ext/MetalExt.jl:10]: I have already opened a PR for this for other backends: https://github.com/JuliaGPU/KernelAbstractions.jl/pull/353

Comment (vchuravy) [reply ext/MetalExt.jl]: Yeah I will try and have that in the 0.9 release

**Comment by vchuravy** [comment]
@tgymnich this now shows the failure that you suspected is due to a synchronization error. cc: @maleadt

**Comment by vchuravy** [comment]
@ChrisRackauckas from my side the big question is can we get rid of the hard dependency on CUDA? I am also not sure what needs to happen for the other tests. Presumably we would need `lufact!` on the other platforms?

File: Project.toml [hunk 3]
```diff
@@ -40,6 +51,8 @@ DocStringExtensions = "0.8, 0.9"
 ForwardDiff = "0.10"
 KernelAbstractions = "0.8"
 LinearSolve = "1.15"
+Metal = "0.2"
+MetalKernels = "0.1"
```
Comment (vchuravy) [inline Project.toml:55]:
```suggestion
MetalKernels = "0.1.1"
```

**Comment by ChrisRackauckas** [comment]
I think it's fine to do a breaking v1.0 to make CUDA.jl no longer a hard dep

**Comment by ChrisRackauckas** [comment]
@utkarsh530 what's required to finish merging the metal stuff?

File: Project.toml [hunk 4]
```diff
@@ -40,6 +51,8 @@ DocStringExtensions = "0.8, 0.9"
 ForwardDiff = "0.10"
 KernelAbstractions = "0.8"
```
Comment (ChrisRackauckas) [inline Project.toml:52]:
```suggestion
KernelAbstractions = "0.9"
```

**Comment by ChrisRackauckas** [comment]
I assume from the release notes this may need KA v0.9?

File: Project.toml (hunk 4 above)
Comment (vchuravy) [reply Project.toml]: No KernelAbstractions 0.9 is a breaking release and will require more work that that

Comment (ChrisRackauckas) [reply Project.toml]: But https://github.com/JuliaGPU/KernelAbstractions.jl/pull/337 is in v0.9?

Comment (vchuravy) [reply Project.toml]: No that was merged onto `release-0.8`

Comment (ChrisRackauckas) [reply Project.toml]: 🤔 https://github.com/JuliaGPU/KernelAbstractions.jl/releases/tag/v0.9.0 okay that's just wrong then.

File: Project.toml [hunk 5]
```diff
@@ -38,8 +49,10 @@ ChainRulesCore = "1"
 DiffEqBase = "6.81"
 DocStringExtensions = "0.8, 0.9"
 ForwardDiff = "0.10"
-KernelAbstractions = "0.8"
+KernelAbstractions = "0.9"
```
Comment (ChrisRackauckas) [inline Project.toml:52]:
```suggestion
KernelAbstractions = "0.8"
```

File: Project.toml (hunk 4 above)
Comment (ChrisRackauckas) [reply Project.toml:52]: Silly tagbot.

**Comment by vchuravy** [comment]
Personally I would wait until KA 0.9 and the backend packages are all ready. The change here should be easy and will clean up a couple of things.

**Comment by utkarsh530** [comment]
Okay, sounds good. I was getting things ready before the submission.
All the tests pass now, so it should be a simple change.

**Comment by utkarsh530** [comment]
@vchuravy, any status?

**Comment by maleadt** [comment]
Ah, I just wasted an hour updating DiffEqGPU only to notice that this otherwise unrelated PR already contains the necessary KA.jl compatibility changes 🤦

@utkarsh530 What's the status of this PR?

**Comment by maleadt** [comment]
> I can reproduce the exception, and locally it doesn't happen on `-g2` either. Annoying. It seems to come from a literal call to `error()` somewhere.

It happens here: https://github.com/SciML/DiffEqGPU.jl/blob/dddcb594ce054c0677bc1b18fdabca2fc0c2eaa9/src/perform_step/gpu_vern7_perform_step.jl#L161

FWIW, DiffEqGPU really shouldn't be calling `error`; KA.jl probably needs to get an output/error/assertion macro.

**Comment by maleadt** [comment]
Looks like these some divergence within the `GPUAV7I` `step!` function:

```
tmp = (45.000000,0.000000)
btilde1 = 0.002547
k1 = (0.000000,-10.000000)
btilde4 = -0.009658
k4 = (-1.633333,-10.000000)
btilde5 = 0.042065
k5 = (-4.555000,-10.000000)
btilde6 = -0.066682
k6 = (-6.095100,-10.000000)
btilde7 = 0.265010
k7 = (-8.840014,-10.000000)
btilde8 = -0.294303
k8 = (-9.250006,-10.000000)
btilde9 = 0.081317
k9 = (-9.999984,-10.000000)
btilde10 = -0.020295
k10 = (-9.999962,-10.000000)
resulting tmp = (-0.000001,-0.000000)
```

vs

```
tmp = (45.000000,0.000000)
btilde1 = 0.002547
k1 = (0.000000,-10.000000)
btilde4 = -0.009658
k4 = (-1.633333,-10.000000)
btilde5 = 0.042065
k5 = (-4.554998,-10.000000)
btilde6 = -0.066682
k6 = (-6.095097,-10.000000)
btilde7 = 0.265010
k7 = (-8.840016,-10.000000)
btilde8 = -0.294303
k8 = (-9.249968,-10.000000)
btilde9 = 0.081317
k9 = (-9.999983,-10.000000)
btilde10 = -0.020295
k10 = (-9.999996,-10.000000)
resulting tmp = (-0.000013,-0.000000)
```

The generated IR/PTX assembly is identical though. Reduced this to the `--device-debug` flag to `ptxas` 🤦

**Comment by maleadt** [comment]
All of the errors are relatively (acceptably?) small though:

```julia
julia> -9.999962f0 ≈ -9.999996f0
true

julia> -9.999984f0 ≈ -9.999983f0
true

julia> -9.250006f0 ≈ -9.249968f0
true

julia> -8.840014f0 ≈ -8.840016f0
true

julia> -6.095100f0 ≈ -6.095097f0
true

julia> -4.555000f0 ≈ -4.554998f0
true
```

**Comment by maleadt** [comment]
> But what seems a bit strange is that everything passes locally, which I am not sure why is that happening.

FWIW, I could reproduce locally using Julia 1.8 and latest master of everything.

> And why the values are changing now?

Yeah, not sure. It probably also depends on the CUDA compiler used, which may have gotten bumped by the CUDA.jl update. I also recently upgraded the CI system's CUDA version.

**Comment by maleadt** [comment]
> Yes, I think the errors seems acceptable, I’ll increase the tolerance and merge it.

The issue is an `error()` thrown from within the kernel; increasing the tolerance won't fix that, right?

> BTW, why does at 1.9-Nightly registry fetches the version Metal.jl 0.2.0 ?

I'm not sure, Metal.jl itself runs 1.9 CI, so it is compatible.

**Comment by utkarsh530** [comment]
I got this for the latest CI test:

https://buildkite.com/julialang/diffeqgpu-dot-jl/builds/677#018777f6-2cf2-4b31-a11e-3e2d9dcbc127/627-1328

**Comment by utkarsh530** [comment]
LGTM now. I tested AMD GPU tests locally, and all passed now.
@ChrisRackauckas let's merge it?

**Pull request merged**